File: groupctl-prototype/groupctl.py
from typing import List
from colors import BITIWG_TO_LAUNCHPAD_COLORS, BwColor, LaunchpadColor
from events import (
    BlinkPad,
    EventBusSubscriber,
    EventBus,
    PageSelected,
    RequestSelectGroup,
    PadClick,
    Event,
    SchemaChanged,
    BwTrackSelected,
    BwTrack,
    LightPadUp,
)
from pager import Page 


def flatten(tracks: List[BwTrack]) -> List[BwTrack]:
    result = []
    for track in tracks:
        result.append(track)
        result.extend(flatten(track.children))
    return result


GLOBAL_TO_LOCAL = {
    29: 13, 30: 14, 31: 15, 32: 16,  # quadrant top row
    21:  9, 22: 10, 23: 11, 24: 12,
    13:  5, 14:  6, 15:  7, 16:  8,
     5:  1,  6:  2,  7:  3,  8:  4,  # quadrant bottom row
}


class LaunchpadGroupCtl(EventBusSubscriber):
    def __init__(self, bus: EventBus) -> None:
        self.bus = bus
        self.bus.subscribe(self)
        self.tracks: List[BwTrack] = []
        self.selected_group_id: str | None = None
        self.page_active: bool = True

    # TODO: cache
    @property
    def groups(self) -> List[BwTrack]:
        without_refs = [
            t for t in self.tracks
            if "top refs" not in t.name
        ]
        return [
            t for t in flatten(without_refs)
            if t.depth in [0, 1] and t.children
        ]

    def _track_id_to_position(self, track_id: str) -> int | None:
        for track in self.groups:
            if track.id == track_id:
                return track.position
        return None

    def _track_position_to_id(self, position: int):
        for track in self.groups:
            if track.position == position:
                return track.id

    def _track_position_to_name(self, position: int):
        for track in self.groups:
            if track.position == position:
                return track.name

    def _global_to_local_position(self, n: int) -> int | None:
        """
        Map a global launchpad pad index (1..64) to a local index (1..16)
        within the bottom-right quadrant. Returns None for pads outside it.
        """
        return GLOBAL_TO_LOCAL.get(n)

    def _local_to_global_position(self, n: int) -> int | None:
        local_to_global = dict([(b, a) for a, b in GLOBAL_TO_LOCAL.items()])
        return local_to_global.get(n)

    def _bw_to_launchpad_color(self, color: BwColor) -> LaunchpadColor | None:
        return BITIWG_TO_LAUNCHPAD_COLORS.get(color)

    def _paint(self) -> None:
        if not self.page_active:
            return
        groups = self.groups
        for gt in groups:
            position = self._local_to_global_position(gt.position)
            color = self._bw_to_launchpad_color(gt.color)
            if position and color:
                event_cls = (
                    BlinkPad
                    if gt.id == self.selected_group_id
                    else LightPadUp
                )
                self.bus.send(event_cls(n=position, color=color))

    def on(self, event: Event) -> None:
        match event:
            case SchemaChanged(tracks=tracks):
                self.tracks = tracks
                self._paint()
            case PadClick(n=n) if self.page_active:
                pos = self._global_to_local_position(n)
                track_id = None
                track_name = None
                if pos:
                    track_id = self._track_position_to_id(pos)
                    track_name = self._track_position_to_name(pos)
                if pos and track_id and track_name:
                    self.bus.send(
                        RequestSelectGroup(
                            track_id=track_id,
                            track_name=track_name,
                        ),
                    )
            case BwTrackSelected(track_id=track_id):
                self.selected_group_id = track_id
                self._paint()
            case PageSelected(n=n):
                self.page_active = n == 0
                self._paint()

File: GroupCtl/src/main/java/dev/tradcode/groupctl/GroupCtl.java
package dev.tradcode.groupctl;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import dev.tradcode.groupctl.events.BitwigTrack;
import dev.tradcode.groupctl.events.BitwigTrackSelected;
import dev.tradcode.groupctl.events.Event;
import dev.tradcode.groupctl.events.IEventBus;
import dev.tradcode.groupctl.events.IEventBusSubscriber;
import dev.tradcode.groupctl.events.PadClicked;
import dev.tradcode.groupctl.events.PageSelected;
import dev.tradcode.groupctl.events.PaintPad;
import dev.tradcode.groupctl.events.RequestSelectGroup;
import dev.tradcode.groupctl.events.SchemaChanged;

public class GroupCtl implements IEventBusSubscriber {
    IEventBus bus;
    ArrayList<BitwigTrack> schema;
    int selectedGroupId = -1;
    boolean pageActive = true; // TODO
    Map<Integer, Integer> GLOBAL_TO_LOCAL = Map.ofEntries(
        // row 1
        Map.entry(15, 1),  Map.entry(16, 2),
        Map.entry(17, 3),  Map.entry(18, 4),
        // row 2
        Map.entry(25, 5),  Map.entry(26, 6),
        Map.entry(27, 7),  Map.entry(28, 8),
        // row 3
        Map.entry(35, 9),  Map.entry(36, 10),
        Map.entry(37, 11), Map.entry(38, 12),
        // row 4
        Map.entry(45, 13), Map.entry(46, 14),
        Map.entry(47, 15), Map.entry(48, 16)
    );

    public GroupCtl(IEventBus bus) {
        this.bus = bus;
        this.bus.subscribe(this);
    }

    private ArrayList<BitwigTrack> flatten(ArrayList<BitwigTrack> tracks) {
        ArrayList<BitwigTrack> res = new ArrayList<>();
        if (tracks.size() == 0) return res;
        for (var t : tracks) {
            res.add(t);
            res.addAll(this.flatten(t.children));
        }
        return res;
    }

    private List<BitwigTrack> getGroups() {
        return this.flatten(this.schema)
            .stream()
            .filter(t -> t.isGroup)
            .toList();
    }

    private int trackIdToPosition(int id) {
        return this.getGroups()
            .stream()
            .filter(t -> t.id == id)
            .findFirst()
            .map(t -> t.getPosition())
            .orElse(-1);
    }

    private int trackPositionToId(int pos) {
        return this.getGroups()
            .stream()
            .filter(t -> t.getPosition() == pos)
            .findFirst()
            .map(t -> t.id)
            .orElse(-1);
    }

    private String trackPositionToName(int pos) {
        return this.getGroups()
            .stream()
            .filter(t -> t.getPosition() == pos)
            .findFirst()
            .map(t -> t.name)
            .orElse(null);
    }

    private int globalToLocalPosition(int gpos) {
        if (gpos == -1) return -1;
        return GLOBAL_TO_LOCAL.get(gpos);
    }

    private int localToGlobalPosition(int pos) {
        if (pos == -1) return -1;
        var reversed = new HashMap<Integer, Integer>();
        GLOBAL_TO_LOCAL.forEach((k, v) -> reversed.put(v, k));
        return reversed.get(pos);
    }

    private int bwToLaunchpadColor(String bwColor) {
        return Colors.toLaunchpad(bwColor);
    }

    private void paint() {
        if (!this.pageActive)
            return;
        for (var gt : this.getGroups()) {
            var pos = this.localToGlobalPosition(gt.getPosition());
            var color = this.bwToLaunchpadColor(gt.color);
            if (pos != -1 && color != -1)
                // TODO: blink distinction
                this.bus.send(
                    new PaintPad(pos, color)
                );
        }
    }

    public void on(Event event) {
        switch (event) {
            case SchemaChanged(ArrayList<BitwigTrack> schema) -> {
                this.schema = schema;
                this.paint();
            }
            case PadClicked(int n) when this.pageActive -> {
                var pos = this.globalToLocalPosition(n);
                if (pos != -1) {
                    var trackId = this.trackPositionToId(n);
                    var trackName = this.trackPositionToName(n);
                    if (trackId != -1 && trackName != null)
                        this.bus.send(
                            new RequestSelectGroup(
                                trackId,
                                trackName
                            )
                        );
                }
            }
            case BitwigTrackSelected(int trackId) -> {
                this.selectedGroupId = trackId;
                this.paint();
            }
            case PageSelected(int n) -> {
                this.pageActive = n == 0;
                this.paint();
            }
            default -> { }
        }
    }
}
